### DISCUSSION TITLE
Order by 'desc' slower then 'asc'

### DISCUSSION THREAD

**Discussion by CyberCyclone** [ravendb/ravendb#13819] [Q&A]
I have an automatic index that I'm doing a test query against that takes 129ms with:

```
from 'Block'
order by block.number as long asc
limit 20
```

If I change it to `desc` it adds 700+ ms to the query time:
```
from 'Block'
order by block.number as long desc
limit 20
```

The raw doc looks something like this and there's 5 million + documents:

```
{
    "context": {
        "block_hash": "3e188b6f0947569324eee92f82d902a35cc5d665c452f680c33212f2d1cb8928",
        "block_number": 5633894,
        "slot": 27739937
    },
    "block": {
        "era": "Mary",
        "body_size": 8341,
        "issuer_vkey": "4dee00c52f783e86fe9bd1fbe78400555b08b5674a8c4661a2696337b9905135",
        "tx_count": 25,
        "slot": 27739937,
        "hash": "3e188b6f0947569324eee92f82d902a35cc5d665c452f680c33212f2d1cb8928",
        "number": 5633894,
        "previous_hash": "233ae5295be909509f4c1188fffd6cc4a7bcdb9e3f625620d2e61d0b2999abcc"
    },
    "timestamp": 1619306228000,
    "@metadata": {
        "@collection": "Block"
    }
}
```

I'm not sure why it would be biased towards `asc` sorting, but is there a way to get more consistent performance? Seeing as `desc` will always be used, is there a way when creating the index to default a certain sort order direction and type (long) so it doesn't have to sort it on every query? I can't seem to find anything relating to it on https://ravendb.net/docs/article-page/5.3/csharp/indexes/sorting-and-collation.

**Comment by ayende** [comment]
This is a very surprising finding. Can you try create a sample data set that we can look at?
There should be no performance difference between `desc` and `asc`. The same rough cost is involved in both

**Reply by CyberCyclone** [reply]
I've tried uploading the collection to [http://live-test.ravendb.net/studio/index.html#databases/tasks/import/file?&database=asc-desc-orderinng-test](http://live-test.ravendb.net/studio/index.html#databases/tasks/import/file?&database=asc-desc-orderinng-test) however it gets about 1/20 of the way and then restarts. After 3 goes it throws an import failed error. It's a 700MB file so I'm guessing that there's a max upload size for the imports.

I'm not sure how else to recreate the data set but here's the RavenDB dump file of just that collection (it's publicly available data): https://1drv.ms/u/s!Ap0D9ohJOI541V__U6UWCHmggf3m?e=xoFsjW

**Comment by CyberCyclone** [comment]
I've just discovered something interesting.

If I add the `as` field, the query is consistent:

With `as` keeps query below 200ms
```
from 'Block' as block
order by block.number as long desc
limit 20
```

Without `as` increases `desc` query to 700ms
```
from 'Block'
order by block.number as long desc
limit 20
```

While it fixes the issue for me, I certainly didn't expect it to make any performance difference. Is this expected behaviour?

**Reply by ayende** [reply]
Okay, that makes a lot more sense, in that case.
The issue is that `as long` means that we can threat the data as numeric, while without it, we have to do string comparisons.
The numeric option is faster, so the difference isn't between `asc` / `desc` but the different sorting options.

Not an issue.

**Reply by CyberCyclone** [reply]
@ayende just to be clear, the `as` I was referring too was after the `from 'Block'` statement. As in, `from 'Block' as block` compared to `from 'Block'`. The `as long` wasn't part of the issue.

**Reply by ayende** [reply]
Oh, I see, I'm sorry for my confusion.

I think that the problem then is that you don't _have_ a field named `'block.number'`, and with the `as block`   it tests for `number`.
Then it can either find the field and optimize or not (and have to do more work)

**Comment by CyberCyclone** [comment]
Hi @ayende I've come across this again and I'm not really sure what's going on.

```
from 'Blocks'
where schemaVersion = 3.5
order by block.number as long 
limit 100
```
^ 121ms

```
from 'Blocks'
where schemaVersion = 3.5
order by block.number as long desc
limit 100
```
^ 708ms

Example document:
```
{
    "timestamp": "2019-11-15T01:43:51Z",
    "schemaVersion": 3.5,
    "block": {
        "era": "Byron",
        "epoch": 156,
        "epoch_slot": 9357,
        "body_size": 671,
        "issuer_vkey": "993a8f056d2d3e50b0ac60139f10df8f8123d5f7c4817b40dac2b5dd8aa94a82e8536832e6312ddfc0787d7b5310c815655ada4fdbcf6b12297d4458eccc2dfb",
        "tx_count": 0,
        "slot": 3378957,
        "hash": "f1ad789157d44fdd1de24e0a2e09b6d4bd27b0c85ce0cd4d25c29f24c75c8cd8",
        "number": 3377417,
        "previous_hash": "ea6f72476532744256a62de4bac0b946556832f66a4ae87b28b01b1dafa452e4",
        "cbor_hex": null,
        "transactions": []
    },
    "@metadata": {
        "@collection": "Blocks",
        "@flags": "FromReplication",
        "@nested-object-types": {},
        "Raven-Node-Type": "Block"
    }
}
```

Currently there's about 3 million documents.

I really don't understand why `desc` would slow the query down. This is with using an auto-index.

**Reply by garayx** [reply]
how do you run the query? if you run it from studio, can you make sure that the cache is disabled in all runs?

**Reply by ayende** [reply]
If you can provide full details (size of the dataset, the index in question, etc - that would help).
Are you running this when there are writes happening in the system?

If this is reproducible on its own, I would love to have a method to see that locally.

There is no reason why this would be the case.
